### PR TITLE
add SameType as default type inference function in imperative mode

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
Add SameType as default type inference function in imperative mode (this is the same way in symbolic mode).

Test locally by removing the `FInferType` attr in Regression operators, before this PR the script below will fail

```
import mxnet as mx
import numpy as np

in_shape = (10, 10, 10)
ctx = mx.cpu()
data = mx.nd.ones(in_shape)
out = mx.nd.LinearRegressionOutput(data=data, label=data)
print out.asnumpy()
```

cc @eric-haibin-lin @reminisce 

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/executor/exec_pass.h [hunk 1]
```diff
@@ -189,6 +189,13 @@ bool DefaultStorageType(const nnvm::NodeAttrs& attrs,
                         std::vector<int> *iattr,
                         std::vector<int> *oattr);
 
+/*! \brief The default type inference function, which assigns all undefined
+ *         types to the same type of one of the inputs or outputs.
+ */
+bool SameType(const nnvm::NodeAttrs& attrs,
```
Comment (piiswrong) [inline src/executor/exec_pass.h]: where is the definition of this function?

Comment (ZiyueHuang) [reply src/executor/exec_pass.h]: I haven’t found the definition of SameType. It seems that this function is just defined and implemented in infer_graph_attr_pass.cc.

A same kind of function, DefaultStorageType, is defined in this header and implemented in infer_graph_attr_pass.cc.

Comment (ZiyueHuang) [reply src/executor/exec_pass.h]: I tried global search SameType in github. It only appears twice in infer_graph_attr_pass.cc https://github.com/apache/incubator-mxnet/search?utf8=✓&q=SameType&type=

File: src/imperative/imperative_utils.h [hunk 2]
```diff
@@ -120,9 +120,13 @@ inline void SetShapeType(const Context& ctx,
   for (auto& i : outputs) {
     out_types.push_back(i->dtype());
   }
-  CHECK(infertype.count(attrs.op))
-    << "Operator " << attrs.op->name << " is missing FInferType attribute";
-  CHECK(infertype[attrs.op](attrs, &in_types, &out_types));
+  bool infer_type_success;
```
Comment (reminisce) [inline src/imperative/imperative_utils.h]: nit: initialize the variable with `false`.

File: src/executor/exec_pass.h (hunk 1 above)
Comment (reminisce) [reply src/executor/exec_pass.h]: Can we move the definition to some file in `src/common` for better code management? It's kind of confusing to declare it here and define it in a file not related to the caller.

Comment (ZiyueHuang) [reply src/executor/exec_pass.h]: Sure. Will update it tomorrow.

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!

**Pull request merged**